### ISSUE TITLE
Using Docker distribution with Mac M2

### ISSUE THREAD

**Issue by psychemedia** [h2oai/h2ogpt#878]
The docs give examples for running containers with NBVIDIA GPU settings.

Is it possible / is there an optimal `docker run` command for running the containers on a Mac M2? Would be useful to have some mention in the docs.

**Comment by kevinschaich** [comment]
+1 here – I was the following error when trying to use the provided `docker-compose.yml` file:

```
98.85 qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
98.85 ./docker_build_script_ubuntu.sh: line 34: python3.10: command not found
------
failed to solve: process "/bin/sh -c cd /workspace && ./docker_build_script_ubuntu.sh" did not complete successfully: exit code: 127
```

I believe this is because images are built for x86/64 Linux but when emulated on Apple Silicon they get run as amd64.

I was able to get the image to build with the following `docker-compose.yml` (add `platform: linux/x86_64` and remove Nvidia-related config):

```
version: '3'

services:
  h2o-llm:
    platform: linux/x86_64
    build:
      context: ./
      dockerfile: Dockerfile
      args:
        BUILDKIT_INLINE_CACHE: "0"
    image: h2o-llm
    shm_size: '64gb'
    command: generate.py --load_8bit True --base_model 'EleutherAI/gpt-j-6B'
    restart: unless-stopped
    volumes:
      - ./h2o-llm:/root/.cache # Location downloaded weights will be stored
      - ./.cache:/workspace/.cache
    ports:
      - 7860:7860
    # deploy:
    #   resources:
    #     reservations:
    #       devices:
    #         - driver: nvidia
    #           count: all
    #           capabilities: [ gpu ]

volumes:
  h2o-llm:
    name: h2o-llm
```

While this was building (very slow, it took ~40 mins) I did some research and I think the native Mac route [described here](https://github.com/h2oai/h2ogpt/blob/main/docs/README_MACOS.md) is the way to go for the following reasons:

- When running with the `platform` arg Docker has to emulate another architecture and it will be massively slower when running operations on the CPU (CPU inference is already slow enough without emulation)
- Can't take full advantage of Docker Virtualization Framework & VirtioFS mount acceleration
- Accelerated GPU training is enabled only when using Apple’s Metal Performance Shaders (MPS) as a backend for PyTorch – for this you have to run natively as Docker probably will not get MPS support any time soon.

TL;DR – unless the authors or community have any creative suggestions, [install it natively](https://github.com/h2oai/h2ogpt/blob/main/docs/README_MACOS.md) or it will probably be too slow for any real-world use cases 🙂

**Comment by psychemedia** [comment]
It's possible to build a native arm46 image, I guess. I'm not sure how that works for enabling any form of acceleration. (I've cross-built non-gpu requiring images before, but handling GPUs in a cross-platform way is perhaps a step too far for a simple cross-build approach.)

**Comment by pseudotensor** [comment]
I'm sure it's possible, but probably not in our roadmap.
